src/ApiClient/util/AuthController.ts

```diff
@@ -0,0 +1,232 @@
+
+    public loggingIn = false;
+
+    public constructor() {
+        if (MODE === "DEV") {
+            window.authController = this;
+        }
+    }
+
+    public Initialize() {
+        this.apiClient = new Api(
+            new HttpClient({
+                //Yes this is only initialized once even if the configOption changes, this doesn't
+                baseURL: configOptions.apipath.value as string,
+                withCredentials: false,
+                headers: {
+                    Accept: "application/json",
+                    Api: `Tgstation.Server.Api/` + API_VERSION,
+                    "Webpanel-Version": VERSION
+                }
+            })
+        );
+    }
+
+    public isTokenValid() {
+        return !!(
+            this.token &&
+            this.token.bearer &&
+            (!this.token.expiresAt || new Date(this.token.expiresAt).getTime() > Date.now())
+        );
+    }
+
+    /**
+     * Get token while handling validation. Returns undefined IF the cached credential is blank.
+     * Will also automatically logout the user.
+     *
+     * "Why not a getter?" because you cant have async getters.
+     */
+    public async getToken() {
+        if (!this.token || !this.isTokenValid()) {
+            if ((await this.authenticateCached()).code !== StatusCode.OK) {
+                return this.token;
+            }
+            this.logout(); // overkill, AuthenticatedCache already runs logout
+            return;
+        }
+        return this.token;
+    }
+
+    /**
+     * Get token without handling validation.
+     */
+    public getTokenUnsafe() {
+        return this.token;
+    }
+
+    /**
+     * Authenticate using the cached credential information
+     */
+    public async authenticateCached() {
+        if (!this._credentials) {
+            return new InternalStatus<undefined, LoginErrors>({
+                code: StatusCode.ERROR,
+                error: new InternalError(ErrorCode.LOGIN_NOCREDS, { void: true })
+            });
+        }
+        return await this.authenticate(this._credentials);
+    }
+
+    /**
+     * Attempt to authenticate using the credentials.
+     * This **will not** return token from status.
+     * @param credential The credentials, either password/username or oauth.
+     * @param cacheCredential Should we cache the credentials for later login?
+     */
+    public async authenticate(credential: ICredentials, cacheCredential = true) {
+        if (!credential) {
+            return new InternalStatus<undefined, LoginErrors>({
+                code: StatusCode.ERROR,
+                error: new InternalError(ErrorCode.LOGIN_NOCREDS, { void: true })
+            });
+        }
+        // debounce (iffy about this)
+        if (this.loggingIn) {
+            return new InternalStatus<undefined, LoginErrors>({
+                code: StatusCode.ERROR,
+                error: new InternalError(ErrorCode.LOGIN_LOGGING_IN, { void: true })
+            });
+        }
```

Comment:
This will break shit, its quite possible for several requests to be queued while waiting for reauthentication

Reply:
fixed